streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImplTest.java

```diff
@@ -613,29 +616,25 @@ public void apply(final String key, final String value) {
     }
 
     private void processData() {
-        driver.setUp(builder, TestUtils.tempDirectory(), Serdes.String(), Serdes.String(), 0);
-        driver.setTime(0);
-        driver.process(TOPIC, "1", "A");
-        driver.process(TOPIC, "2", "B");
-        driver.process(TOPIC, "1", "C");
-        driver.process(TOPIC, "1", "D");
-        driver.process(TOPIC, "3", "E");
-        driver.process(TOPIC, "3", "F");
-        driver.process(TOPIC, "3", null);
-        driver.flushState();
+        driver = new TopologyTestDriver(builder.build(), props);
+        driver.pipeInput(recordFactory.create(TOPIC, "1", "A"));
+        driver.pipeInput(recordFactory.create(TOPIC, "2", "B"));
+        driver.pipeInput(recordFactory.create(TOPIC, "1", "C"));
+        driver.pipeInput(recordFactory.create(TOPIC, "1", "D"));
+        driver.pipeInput(recordFactory.create(TOPIC, "3", "E"));
+        driver.pipeInput(recordFactory.create(TOPIC, "3", "F"));
+        driver.pipeInput(recordFactory.create(TOPIC, "3", null));
     }
 
     private void doCountWindowed(final List<KeyValue<Windowed<String>, Long>> results) {
-        driver.setUp(builder, TestUtils.tempDirectory(), 0);
-        driver.setTime(0);
-        driver.process(TOPIC, "1", "A");
-        driver.process(TOPIC, "2", "B");
-        driver.process(TOPIC, "3", "C");
-        driver.setTime(500);
-        driver.process(TOPIC, "1", "A");
-        driver.process(TOPIC, "1", "A");
-        driver.process(TOPIC, "2", "B");
-        driver.process(TOPIC, "2", "B");
+        driver = new TopologyTestDriver(builder.build(), props);
+        driver.pipeInput(recordFactory.create(TOPIC, "1", "A", 0));
+        driver.pipeInput(recordFactory.create(TOPIC, "2", "B", 0));
+        driver.pipeInput(recordFactory.create(TOPIC, "3", "C", 0));
+        driver.pipeInput(recordFactory.create(TOPIC, "1", "A", 500));
```

Comment:
Might be simpler to call `recordFactory.advanceTimeMs()` and not specify the timestamp for each record explicitly? We also leave it as is.

Reply:
Because in the remaining tests in this class we are also passing the timestamp, I'd like to keep it coherent and use the same here. But I can change it if you see advantages in calling `recordFactory.advanceTimeMs()`

Reply:
Ack. Fine to leave as is -- was just a thought.